File: Source/NextTurnRuntime/Private/Slate/GridPanelSlot.cpp
// Copyright (c) NextTurn. All rights reserved.
// Licensed under the Apache License, Version 2.0.
// See LICENSE.txt in the project root for more information.

#if WITH_APPLICATION_CORE

#include "NextTurnRuntime/Private/Exports.hpp"

#include "NextTurnRuntime/Public/Compiler.hpp"

#include "Runtime/Core/Public/CoreMinimal.h"

#include "Runtime/Slate/Public/Widgets/Layout/SGridPanel.h"

ENABLE_WARNINGS;

#define TYPE_EXPORT_MANAGED_TYPE_NAME GridPanelSlot

TYPE_EXPORT_START

void EXPORT_CALL_CONV SetColumn(SGridPanel::FSlot& Slot, int32_t Column)
{
	Slot.Column(Column);
}

void EXPORT_CALL_CONV SetColumnSpan(SGridPanel::FSlot& Slot, int32_t ColumnSpan)
{
	Slot.ColumnSpan(ColumnSpan);
}

void EXPORT_CALL_CONV SetRow(SGridPanel::FSlot& Slot, int32_t Row)
{
	Slot.Row(Row);
}

void EXPORT_CALL_CONV SetRowSpan(SGridPanel::FSlot& Slot, int32_t RowSpan)
{
	Slot.RowSpan(RowSpan);
}

const FMemberSymbol Members[] =
{
	EXPORT_METHOD(SetColumn),
	EXPORT_METHOD(SetColumnSpan),
	EXPORT_METHOD(SetRow),
	EXPORT_METHOD(SetRowSpan),
};

TYPE_EXPORT_END

#undef TYPE_EXPORT_MANAGED_TYPE_NAME

RESTORE_WARNINGS;

#endif

File: Managed/NextTurn.UE.Runtime/Slate/GridPanelSlot.cs
// Copyright (c) NextTurn. All rights reserved.
// Licensed under the Apache License, Version 2.0.
// See LICENSE.txt in the project root for more information.

using NextTurn.UE.Annotations;

namespace Unreal.Slate
{
    public class GridPanelSlot : Slot
    {
        private unsafe NativeGridPanelSlot* Target => (NativeGridPanelSlot*)this.pointer;

        public unsafe int Column
        {
            get => this.Target->Column;
            set => NativeMethods.SetColumn(this.Target, value);
        }

        public unsafe int ColumnSpan
        {
            get => this.Target->ColumnSpan;
            set => NativeMethods.SetColumnSpan(this.Target, value);
        }

        public unsafe int Row
        {
            get => this.Target->Row;
            set => NativeMethods.SetRow(this.Target, value);
        }

        public unsafe int RowSpan
        {
            get => this.Target->RowSpan;
            set => NativeMethods.SetRowSpan(this.Target, value);
        }

        private struct NativeGridPanelSlot
        {
            public int Column { get; }

            public int ColumnSpan { get; }

            public int Row { get; }

            public int RowSpan { get; }
        }

        private static class NativeMethods
        {
            [Calli]
            public static extern unsafe void SetColumn(NativeGridPanelSlot* slot, int column);

            [Calli]
            public static extern unsafe void SetColumnSpan(NativeGridPanelSlot* slot, int columnSpan);

            [Calli]
            public static extern unsafe void SetRow(NativeGridPanelSlot* slot, int row);

            [Calli]
            public static extern unsafe void SetRowSpan(NativeGridPanelSlot* slot, int rowSpan);
        }
    }
}
